docs/DeveloperGuide.md

```diff
@@ -451,6 +453,72 @@ Step 10. The headers and rows are written to the CSV file that is specified by t
 
 <div style="page-break-before: always;"></div>
 
+### Command History feature
+
+#### Implementation
+
+The operation is exposed in `CommandBox` and `CommandHistory`.
+
+The user keystroke will be read in `CommandBox#handleKeyStroke`, which will call the appropriate method in `CommandHistory` to retrieve the previous and next commands. The retrieved commands are passed into `CommandBox#commandTextField` to be displayed.
+
+Given below is an example usage scenario and how the Find mechanism behaves at each step.
+
+Step 1. The user launches the application for the first time. All contacts are displayed by default.
+
+Step 2. The user executes `find n/David t/friend t/football` to search for a matching contact.
+
+Step 3. User wants to retrieve previous command and presses the `UP` key.
+
+Step 4. `CommandBox#handleKeyStroke` reads the `UP` keystroke and calls `CommandHistory#getPreviousCommand`.
+
+Step 5. `CommandHistory` retrieves the previous command and returns it.
+
+Step 6. `CommandBox` displays the previous command in the Command Box.
+
+#### Design considerations:
+
+**Aspect: Implementation of Command History:**
+
+* **Option 1 (current choice):** Generate `CommandHistory` as a Singleton class.
+    * Pros: Ensures that there is only one set of history that is tracked.
+    * Cons: Complicates testing as different versions of history cannot be created.
+
+* **Option 2:** Generate `CommandHistory` as a normal class.
+    * Pros: Makes testing easier as multiple `CommandHistory` objects can be created with different history to test different conditions.
+    * Cons: Possible conflicting history if `CommandHistory` is not updated properly.
+
+### Command Assistant feature
+
+#### Implementation
+
+The operation is exposed in `SystemCommand`, specifically in `SystemCommand#execute`.
+
+The user keystroke will be read in `CommandBox#handleKeyStroke`. The user input will be evaluated by `SystemCommand#execute` to determine the appropriate help message to display.
+
+Given below is an example usage scenario and how the Find mechanism behaves at each step.
+
+Step 1. The user launches the application for the first time. All contacts are displayed by default.
+
+Step 2. The user enters `find ` to search for a matching contact.
+
+Step 3. `CommandBox#handleKeyStroke` reads `find ` keystroke and calls `SystemCommand#execute`.
```

Comment:
is `find ` a keystroke or a String?
Perhaps the space is a keystroke but it reads the String